File: resolver/src/main/java/com/leo/sp/resolver/SpResolver.java
package com.leo.sp.resolver;

import android.content.ContentResolver;
import android.content.ContentUris;
import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
/**
 * Created by LEO
 * On 2019/10/13
 */
public class SpResolver {
    private Context mContext;
    private String mAuthority;

    public SpResolver(Context mContext, String authority) {
        this.mContext = mContext;
        this.mAuthority = authority;
    }

    /**
     * int
     *
     * @param key
     * @return 默认返回值是0
     */
    public int getInt(String key) {
        int result = 0;
        Uri uri = ContentUris.withAppendedId(getUri(), SpConfigParams.INT);
        Cursor cursor = mContext.getContentResolver().query(uri, null, key, null, null);
        if (null != cursor) {
            if (cursor.moveToNext()) {
                result = cursor.getInt(cursor.getColumnIndex("data"));
            }
            if (!cursor.isClosed()) {
                cursor.close();
            }
        }
        return result;
    }

    public void setInt(String key, int value) {
        Uri uri = ContentUris.withAppendedId(getUri(), SpConfigParams.INT);
        ContentValues values = new ContentValues();
        values.put(key, value);
        ContentResolver contentResolver = mContext.getContentResolver();
        contentResolver.insert(uri, values);
        contentResolver.notifyChange(uri, null);
    }

    /**
     * long
     *
     * @param key
     * @return 默认返回值0
     */
    public long getLong(String key) {
        long result = 0L;
        Uri uri = ContentUris.withAppendedId(getUri(), SpConfigParams.LONE);
        Cursor cursor = mContext.getContentResolver().query(uri, null, key, null, null);
        if (null != cursor) {
            if (cursor.moveToNext()) {
                result = cursor.getLong(cursor.getColumnIndex("data"));
            }
            if (!cursor.isClosed()) {
                cursor.close();
            }
        }
        return result;
    }

    public void setLong(String key, long value) {
        Uri uri = ContentUris.withAppendedId(getUri(), SpConfigParams.LONE);
        ContentValues values = new ContentValues();
        values.put(key, value);
        ContentResolver contentResolver = mContext.getContentResolver();
        contentResolver.insert(uri, values);
        contentResolver.notifyChange(uri, null);
    }

    /**
     * float
     *
     * @param key
     * @return 默认返回值0
     */
    public float getFloat(String key) {
        float result = 0;
        Uri uri = ContentUris.withAppendedId(getUri(), SpConfigParams.FLOAT);
        Cursor cursor = mContext.getContentResolver().query(uri, null, key, null, null);
        if (null != cursor) {
            if (cursor.moveToNext()) {
                result = cursor.getFloat(cursor.getColumnIndex("data"));
            }
            if (!cursor.isClosed()) {
                cursor.close();
            }
        }
        return result;
    }

    public void setFloat(String key, float value) {
        Uri uri = ContentUris.withAppendedId(getUri(), SpConfigParams.FLOAT);
        ContentValues values = new ContentValues();
        values.put(key, value);
        ContentResolver contentResolver = mContext.getContentResolver();
        contentResolver.insert(uri, values);
        contentResolver.notifyChange(uri, null);
    }

    /**
     * boolean
     * 用int来实现boolean
     *
     * @param key
     * @return 默认返回值false
     */
    public boolean getBoolean(String key) {
        boolean result = false;
        Uri uri = ContentUris.withAppendedId(getUri(), SpConfigParams.INT);
        Cursor cursor = mContext.getContentResolver().query(uri, null, key, null, null);
        if (null != cursor) {
            if (cursor.moveToNext()) {
                int b = cursor.getInt(cursor.getColumnIndex("data"));
                result = b == 1;
            }
            if (!cursor.isClosed()) {
                cursor.close();
            }
        }
        return result;
    }

    public void setBoolean(String key, boolean value) {
        Uri uri = ContentUris.withAppendedId(getUri(), SpConfigParams.INT);
        ContentValues values = new ContentValues();
        values.put(key, value ? 1 : 0);
        ContentResolver contentResolver = mContext.getContentResolver();
        contentResolver.insert(uri, values);
        contentResolver.notifyChange(uri, null);
    }

    /**
     * String
     *
     * @param key
     * @return
     */
    public String getString(String key) {
        String result = "";
        Uri uri = ContentUris.withAppendedId(getUri(), SpConfigParams.STRING);
        Cursor cursor = mContext.getContentResolver().query(uri, null, key, null, null);
        if (null != cursor) {
            if (cursor.moveToNext()) {
                result = cursor.getString(cursor.getColumnIndex("data"));
            }
            if (!cursor.isClosed()) {
                cursor.close();
            }
        }
        return result;
    }

    public void setString(String key, String value) {
        Uri uri = ContentUris.withAppendedId(getUri(), SpConfigParams.STRING);
        ContentValues values = new ContentValues();
        values.put(key, value);
        ContentResolver contentResolver = mContext.getContentResolver();
        contentResolver.insert(uri, values);
        contentResolver.notifyChange(uri, null);
    }

    private Uri getUri() {
        String url = SpConfigParams.PREFIX + SpConfigParams.prefix_suffix
                + mAuthority + SpConfigParams.separator + SpConfigParams.PATH;
        return Uri.parse(url);
    }
}

File: resolver/src/main/java/com/leo/sp/resolver/SpResolver.kt
package com.leo.sp.resolver

import android.content.ContentValues
import android.content.Context
import android.net.Uri
import java.lang.ref.WeakReference

class SpResolver private constructor() {
    private var contextWeakReference: WeakReference<Context>? = null
    private val context: Context
        get() {
            checkNotNull(contextWeakReference) { "context has not been initialed" }
            return contextWeakReference!!.get()
                ?: throw IllegalStateException("context has not been initialed")
        }

    fun init(mContext: Context) {
        contextWeakReference = WeakReference(mContext)
    }

    fun putBoolean(name: String, t: Boolean) {
        val context = context
        val cr = context.contentResolver
        synchronized(mLock) {
            val uri = Uri.parse(
                SpContants.CONTENT_URI + SpContants.SEPARATOR
                        + SpContants.TYPE_BOOLEAN + SpContants.SEPARATOR + name
            )
            val cv = ContentValues()
            cv.put(SpContants.VALUE, t)
            cr.update(uri, cv, null, null)
        }
    }

    fun putString(name: String, t: String) {
        val context = context
        val cr = context.contentResolver
        synchronized(mLock) {
            val uri = Uri.parse(
                SpContants.CONTENT_URI + SpContants.SEPARATOR
                        + SpContants.TYPE_STRING + SpContants.SEPARATOR + name
            )
            val cv = ContentValues()
            cv.put(SpContants.VALUE, t)
            cr.update(uri, cv, null, null)
        }
    }

    fun putInt(name: String, t: Int) {
        val context = context
        val cr = context.contentResolver
        synchronized(mLock) {
            val uri = Uri.parse(
                SpContants.CONTENT_URI + SpContants.SEPARATOR
                        + SpContants.TYPE_INT + SpContants.SEPARATOR + name
            )
            val cv = ContentValues()
            cv.put(SpContants.VALUE, t)
            cr.update(uri, cv, null, null)
        }
    }

    fun putLong(name: String, t: Long) {
        val context = context
        val cr = context.contentResolver
        synchronized(mLock) {
            val uri = Uri.parse(
                SpContants.CONTENT_URI + SpContants.SEPARATOR
                        + SpContants.TYPE_LONG + SpContants.SEPARATOR + name
            )
            val cv = ContentValues()
            cv.put(SpContants.VALUE, t)
            cr.update(uri, cv, null, null)
        }
    }

    fun putFloat(name: String, t: Float) {
        val context = context
        val cr = context.contentResolver
        synchronized(mLock) {
            val uri = Uri.parse(
                SpContants.CONTENT_URI + SpContants.SEPARATOR
                        + SpContants.TYPE_FLOAT + SpContants.SEPARATOR + name
            )
            val cv = ContentValues()
            cv.put(SpContants.VALUE, t)
            cr.update(uri, cv, null, null)
        }
    }

    fun putDouble(name: String, t: Double) {
        val context = context
        val cr = context.contentResolver
        synchronized(mLock) {
            val uri = Uri.parse(
                SpContants.CONTENT_URI + SpContants.SEPARATOR
                        + SpContants.TYPE_DOUBLE + SpContants.SEPARATOR + name
            )
            val cv = ContentValues()
            cv.put(SpContants.VALUE, t)
            cr.update(uri, cv, null, null)
        }
    }

    fun getString(name: String, defaultValue: String): String {
        val context = context
        val cr = context.contentResolver
        val uri = Uri.parse(
            SpContants.CONTENT_URI + SpContants.SEPARATOR
                    + SpContants.TYPE_STRING + SpContants.SEPARATOR + name
        )
        val rtn = cr.getType(uri)
        return if (isInvalidResult(rtn)) {
            defaultValue
        } else rtn!!
    }

    fun getInt(name: String, defaultValue: Int): Int {
        val context = context
        val cr = context.contentResolver
        val uri = Uri.parse(
            SpContants.CONTENT_URI + SpContants.SEPARATOR
                    + SpContants.TYPE_INT + SpContants.SEPARATOR + name
        )
        val rtn = cr.getType(uri)
        return if (isInvalidResult(rtn)) {
            defaultValue
        } else rtn!!.toInt()
    }

    fun getFloat(name: String, defaultValue: Float): Float {
        val context = context
        val cr = context.contentResolver
        val uri = Uri.parse(
            SpContants.CONTENT_URI + SpContants.SEPARATOR
                    + SpContants.TYPE_FLOAT + SpContants.SEPARATOR + name
        )
        val rtn = cr.getType(uri)
        return if (isInvalidResult(rtn)) {
            defaultValue
        } else rtn!!.toFloat()
    }

    fun getDouble(name: String, defaultValue: Double): Double {
        val context = context
        val cr = context.contentResolver
        val uri = Uri.parse(
            SpContants.CONTENT_URI + SpContants.SEPARATOR
                    + SpContants.TYPE_DOUBLE + SpContants.SEPARATOR + name
        )
        val rtn = cr.getType(uri)
        return if (isInvalidResult(rtn)) {
            defaultValue
        } else rtn!!.toDouble()
    }

    fun getBoolean(name: String, defaultValue: Boolean): Boolean {
        val context = context
        val cr = context.contentResolver
        val uri = Uri.parse(
            SpContants.CONTENT_URI + SpContants.SEPARATOR
                    + SpContants.TYPE_BOOLEAN + SpContants.SEPARATOR + name
        )
        val rtn = cr.getType(uri)
        return if (isInvalidResult(rtn)) {
            defaultValue
        } else {
            rtn.toBoolean()
        }
    }

    fun getLong(name: String, defaultValue: Long): Long {
        val context = context
        val cr = context.contentResolver
        val uri = Uri.parse(
            SpContants.CONTENT_URI + SpContants.SEPARATOR
                    + SpContants.TYPE_LONG + SpContants.SEPARATOR + name
        )
        val rtn = cr.getType(uri)
        return if (isInvalidResult(rtn)) {
            defaultValue
        } else rtn!!.toLong()
    }

    operator fun contains(name: String): Boolean {
        val context = context
        val cr = context.contentResolver
        val uri = Uri.parse(
            SpContants.CONTENT_URI + SpContants.SEPARATOR
                    + SpContants.TYPE_CONTAIN + SpContants.SEPARATOR + name
        )
        val rtn = cr.getType(uri)
        return if (isInvalidResult(rtn)) {
            false
        } else {
            rtn.toBoolean()
        }
    }

    /**
     *
     * @param name String
     * @param type String {@link SpContants#TYPE_STRING,SpContants#TYPE_INT,SpContants#TYPE_LONG,
     *                     SpContants#TYPE_FLOAT,SpContants#TYPE_DOUBLE,SpContants#TYPE_BOOLEAN}
     */
    fun remove(name: String, @ContentType type: String) {
        val context = context
        val cr = context.contentResolver
        val uri = Uri.parse(
            SpContants.CONTENT_URI + SpContants.SEPARATOR
                    + type + SpContants.SEPARATOR + name
        )
        cr.delete(uri, null, null)
    }

    fun clear() {
        val context = context
        val cr = context.contentResolver
        val uri = Uri.parse(SpContants.CONTENT_URI + SpContants.SEPARATOR + SpContants.TYPE_CLEAN)
        cr.delete(uri, null, null)
    }

    private fun isInvalidResult(result: String?): Boolean {
        return result == null || result == SpContants.NULL_STRING;
    }

    companion object {
        @Volatile
        private var mInstance: SpResolver? = null
        private val mLock = Any()

        fun getInstance(): SpResolver {
            return mInstance ?: synchronized(this) {
                mInstance ?: SpResolver().also { mInstance = it }
            }
        }
    }
}